subxt/examples/unstable_light_client_parachains.rs

```diff
@@ -0,0 +1,109 @@
+const ASSET_HUB_SPEC: &str =
+    include_str!("../../artifacts/demo_chain_specs/polkadot_asset_hub.json");
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // The smoldot logs are informative:
+    tracing_subscriber::fmt::init();
+
+    // Connecting to a parachain is a multi step process.
+
+    // Step 1. Construct a new smoldot client and subxt builder.
+    let mut client = subxt_lightclient::Client::new(subxt_lightclient::DefaultPlatform::new(
+        "subxt-example-light-client".into(),
+        "version-0".into(),
+    ));
+    let mut builder = LightClient::raw_builder();
+
+    // Step 2. Connect to the relay chain of the parachain. For this example, the Polkadot relay chain.
+    let polkadot_connection = client
+        .add_chain(subxt_lightclient::AddChainConfig {
+            specification: POLKADOT_SPEC,
+            json_rpc: subxt_lightclient::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: NonZeroU32::new(128).unwrap(),
+                max_subscriptions: 1024,
+            },
+            potential_relay_chains: iter::empty(),
+            database_content: "",
+            user_data: (),
+        })
+        .expect("Light client chain added with valid spec; qed");
+    let polkadot_json_rpc_responses = polkadot_connection
+        .json_rpc_responses
+        .expect("Light client configured with json rpc enabled; qed");
+    let polkadot_chain_id = polkadot_connection.chain_id;
+
+    // Add the chain to the light client builder.
+    builder.add_chain(polkadot_chain_id, polkadot_json_rpc_responses);
+
+    println!("Added Polkadot relay chain");
+
+    // Step 3. Connect to the parachain. For this example, the Asset hub parachain.
+    let assethub_connection = client
+        .add_chain(subxt_lightclient::AddChainConfig {
+            specification: ASSET_HUB_SPEC,
+            json_rpc: subxt_lightclient::AddChainConfigJsonRpc::Enabled {
+                max_pending_requests: NonZeroU32::new(128).unwrap(),
+                max_subscriptions: 1024,
+            },
+            // The chain specification of the asset hub parachain mentions that the identifier
+            // of its relay chain is `polkadot`.
+            potential_relay_chains: [polkadot_chain_id].into_iter(),
+            database_content: "",
+            user_data: (),
+        })
+        .expect("Light client chain added with valid spec; qed");
+    let parachain_json_rpc_responses = assethub_connection
+        .json_rpc_responses
+        .expect("Light client configured with json rpc enabled; qed");
+    let parachain_chain_id = assethub_connection.chain_id;
+
+    // Add the chain to the light client builder.
+    builder.add_chain(parachain_chain_id, parachain_json_rpc_responses);
+
+    println!("Added AssetHub parachain");
+
+    // Step 4. Turn the smoldot client into a subxt light client using the builder.
+    let polkadot_api = builder
```

Comment:
As a nit, I think I'd find the example a little easier to follow if all of the builder stuff was in one place, like:

```rust
let polkadot_api = builder
    .add_chain(polkadot_chain_id, polkadot_json_rpc_responses)
    .add_chain(parachain_chain_id, parachain_json_rpc_responses)
    .build(client, polkadot_chain_id)
    .await?;
```

I guess that `client` is used to make requests, but we also need to hold onto those `json_rpc_responses` for each chain, but looking at this I do wonder if it's possible to do any better here, eg avoid those `.add_chain` steps because that information can be extracted from the `client` on the `.build` step or something (I guess I'll know better when I read the rest of the PR :D)

Reply:
`Client` can be indexed on a chain Id for instance to get some stuff (I didn't check what); I wonder if that helps at all:

https://docs.rs/smoldot-light/latest/src/smoldot_light/lib.rs.html#1097

Reply:
I believe regarding the json_rpc_responses object, if the user drops this object, than the backend might not provide any responses:

```rust
pub struct JsonRpcResponses {
    /// Receiving side for responses.
    ///
    /// As long as this object is alive, the JSON-RPC service will continue running. In order
    /// to prevent that from happening, we destroy it as soon as the
    /// [`JsonRpcResponses::public_api_chain_destroyed`] is notified of the destruction of
    /// the sender.
    inner: Option<json_rpc_service::Frontend>,

    /// Notified when the [`PublicApiChain`] is destroyed.
    public_api_chain_destroyed: pin::Pin<Box<event_listener::EventListener>>,
}
```

Extracted from [smolodot](https://github.com/smol-dot/smoldot/blob/c5ec15dd8279f08796fa2271221ad061ef77f381/light-base/src/lib.rs#L306-L310).

Wrt https://docs.rs/smoldot-light/latest/src/smoldot_light/lib.rs.html#1097, I believe that returns the user data that we provided as `()` for default purposes. I couldn't dig anything that's related to the frontend channel that gets responses back